### PR TITLE
[Model] Transformer CN comment.

### PR DESCRIPTION
**Pull request by NANWOOD**
## Description

Transformer CN comment.
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the `lint.sh` to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] Code is well-documented.
- [ ] Related issue is referred in this PR.

### CODE REVIEW

File: llama/model_cn_comment.py [hunk 1]
```diff
@@ -480,7 +493,9 @@ def __init__(self, params: ModelArgs):
             params.dim, params.vocab_size, bias=False, init_method=lambda x: x
         )
 
+        # 计算用于 RoPE 的复指数频率张量.
         self.freqs_cis = precompute_freqs_cis(
+            # ? 暂时搁置
```
Comment (keli-wen) [inline llama/model_cn_comment.py:498]: File an issue and add a [TODO] tag to refer this issue.

File: llama/model_cn_comment.py [hunk 2]
```diff
@@ -490,39 +505,51 @@ def __init__(self, params: ModelArgs):
 
     @torch.inference_mode()
     def forward(self, tokens: torch.Tensor, start_pos: int):
-        """Perform a forward pass through the Transformer model.
+        """执行 Transformer 模型的前向过程.
```
Comment (keli-wen) [inline llama/model_cn_comment.py:508]:
```suggestion
        """执行 Transformer 模型的前向传播.
```

File: llama/model_cn_comment.py [hunk 3]
```diff
@@ -490,39 +505,51 @@ def __init__(self, params: ModelArgs):
...
+            tokens (torch.Tensor): 输入的 token 索引.
+            start_pos (int): Attention caching 的起始位置.
 
         Returns:
-            torch.Tensor: Output logits after applying the Transformer model.
+            torch.Tensor: Transformer 模型输出的 logits.
 
         """
         _bsz, seqlen = tokens.shape
+        # 根据输入的 token 索引计算对应的 embedding
```
Comment (keli-wen) [inline llama/model_cn_comment.py:519]:
```suggestion
        # 根据输入的 token 索引计算对应的 embedding.
```

Comment (keli-wen) [reply llama/model_cn_comment.py:519]: Miss the `[Shape]` comment.

File: llama/model_cn_comment.py [hunk 4]
```diff
@@ -490,39 +505,51 @@ def __init__(self, params: ModelArgs):
...
+        # 若序列长度大于 1, 则需要计算用于 Attention 的 mask.
         if seqlen > 1:
+            # 首先根据序列长度构建一个 [seqlen, seqlen] 的矩阵, 每一个元素的值均为 -inf.
             mask = torch.full(
                 (seqlen, seqlen), float("-inf"), device=tokens.device
             )
 
+            # 保留该矩阵的上三角部分, 不保留主对角线的值, 即主对角线右上部分的元素保持为 -inf,
+            # 其余元素均为0.
+            # 表示每个 token 只能使用当前 token 以及已经处理过的 token 的信息.
```
Comment (keli-wen) [inline llama/model_cn_comment.py:535]: How about add a simple matrix here for clarity.
```
# e.g. mask matrix:
# ┌         ┐
# │ 0 -∞ -∞ │
# │ 0  0 -∞ │
# │ 0  0  0 │
# └         ┘
```

File: llama/model_cn_comment.py [hunk 5]
```diff
@@ -490,39 +505,51 @@ def __init__(self, params: ModelArgs):
...
             mask = torch.triu(mask, diagonal=1)
 
-            # When performing key-value caching, we compute the attention scores
-            # only for the new sequence. Thus, the matrix of scores is of size
-            # (seqlen, cache_len + seqlen), and the only masked entries are (i, j) for
-            # j > cache_len + i, since row i corresponds to token cache_len + i.
+            # 由于使用了 key-value caching, 我们只需要为新序列计算 attention scores.
```
Comment (keli-wen) [inline llama/model_cn_comment.py:538]: 💎💎💎The statement need to refine. And I think add some newlines will have better read experience.

**Review by keli-wen — CHANGES_REQUESTED** [review]

File: llama/model_cn_comment.py [hunk 6]
```diff
@@ -260,63 +262,71 @@ def forward(
...
 
+        # 调整 query, key, value 的尺寸, 完成特征在多头上的分配.
         xq = xq.view(bsz, seqlen, self.n_local_heads, self.head_dim)
         xk = xk.view(bsz, seqlen, self.n_local_kv_heads, self.head_dim)
         xv = xv.view(bsz, seqlen, self.n_local_kv_heads, self.head_dim)
 
+        # 对 query 和 key 执行 RoPE 操作.
         xq, xk = apply_rotary_emb(xq, xk, freqs_cis=freqs_cis)
 
+        # 取出 KV-Cache 中 key 和 value 的缓存.
```
[the remark this answers is not in the document]
Comment (NANWOOD) [reply llama/model_cn_comment.py:289]: Just typos.
```suggestion
```

File: llama/model_cn_comment.py [hunk 7]
```diff
@@ -260,63 +262,71 @@ def forward(
...
-        )  # (bs, cache_len + seqlen, n_local_heads, head_dim)
-
-        xq = xq.transpose(1, 2)  # (bs, n_local_heads, seqlen, head_dim)
-        keys = keys.transpose(
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
-        values = values.transpose(
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
+        # 若 n_kv_heads < n_heads, 则在 n_kv_heads 维度上进行重复扩充
```
[the remark this answers is not in the document]
Comment (NANWOOD) [reply llama/model_cn_comment.py:301]: Seems necessary, I will do it later.

File: llama/model_cn_comment.py [hunk 8]
```diff
@@ -260,63 +262,71 @@ def forward(
...
+        # [Shape] keys: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        keys = keys.transpose(1, 2)
+        # [Shape] values: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        values = values.transpose(1, 2)
+        # 计算 attention scores.
         scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(
             self.head_dim
         )
+        # 若存在 mask, 则对 attention scores 执行加法操作将 masked 部分置为 -inf.
+        # [Shape] scores: (batch_size, n_local_heads, seqlen, cache_len + seqlen)
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:319]:
```suggestion
```

File: llama/model_cn_comment.py [hunk 9]
```diff
@@ -260,63 +262,71 @@ def forward(
...
+        xq = xq.transpose(1, 2)
+        # [Shape] keys: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        keys = keys.transpose(1, 2)
+        # [Shape] values: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        values = values.transpose(1, 2)
+        # 计算 attention scores.
         scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(
             self.head_dim
         )
+        # 若存在 mask, 则对 attention scores 执行加法操作将 masked 部分置为 -inf.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:318]:
```suggestion
        # [Shape] scores: (batch_size, n_local_heads, seqlen, cache_len + seqlen)
        # 若存在 mask, 则对 attention scores 执行加法操作将 masked 部分置为 -inf.
```

File: llama/model_cn_comment.py [hunk 10]
```diff
@@ -330,24 +340,24 @@ def __init__(
...
+            w3 (ColumnParallelLinear): 第三层的线性变换.
 
         """
         super().__init__()
         hidden_dim = int(2 * hidden_dim / 3)
-        # custom dim factor multiplier
+        # 自定义的隐藏层维度缩放参数.
         if ffn_dim_multiplier is not None:
             hidden_dim = int(ffn_dim_multiplier * hidden_dim)
         hidden_dim = multiple_of * (
```
[the remark this answers is not in the document]
Comment (NANWOOD) [reply llama/model_cn_comment.py:363]: Seems this is already stated in the args section.

File: llama/model_cn_comment.py [hunk 11]
```diff
@@ -384,21 +394,21 @@ class TransformerBlock(nn.Module):
...
-            feed_forward (FeedForward): FeedForward module.
-            layer_id (int): Identifier for the layer.
-            attention_norm (RMSNorm): Layer normalization for attention output.
-            ffn_norm (RMSNorm): Layer normalization for feedforward output.
+            n_heads (int): Attention head 的数量.
+            dim (int): 模块的特征维度.
+            head_dim (int): 每一个 attention head 的特征维度.
+            attention (Attention): Attention 模块.
+            feed_forward (FeedForward): FeedForward 模块.
+            layer_id (int): 网络层的标记符.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:409]:
```suggestion
            layer_id (int): 网络层的标识符.
```

File: llama/model_cn_comment.py [hunk 12]
```diff
@@ -423,21 +433,23 @@ def forward(
         freqs_cis: torch.Tensor,
         mask: Optional[torch.Tensor],
     ):
-        """Perform a forward pass through the TransformerBlock.
+        """执行 TransformerBlock 的前向过程.
```
[the remark this answers is not in the document]
Comment (NANWOOD) [reply llama/model_cn_comment.py:436]: Seems not necessary.

File: llama/model_cn_comment.py [hunk 13]
```diff
@@ -423,21 +433,23 @@ def forward(
...
 
         Args:
-            x (torch.Tensor): Input tensor.
-            start_pos (int): Starting position for attention caching.
-            freqs_cis (torch.Tensor): Precomputed cosine and sine frequencies.
-            mask (torch.Tensor, optional): Masking tensor for attention. Defaults to None.
+            x (torch.Tensor): 输入张量.
+            start_pos (int): Attention caching 的起始位置.
+            freqs_cis (torch.Tensor): 预计算的复指数频率张量.
+            mask (torch.Tensor, optional): 计算 attention 时的遮罩. 默认设为 None.
```
[the remark this answers is not in the document]
Comment (NANWOOD) [reply llama/model_cn_comment.py:442]:
```suggestion
            mask (torch.Tensor, optional): 计算 attention 时的 mask. 默认设为 None.
```
"遮罩" seems official translation.

File: llama/model_cn_comment.py (hunk 4 above)
Comment (NANWOOD) [inline llama/model_cn_comment.py:535]:
```suggestion
            # 表示每个 token 只能使用当前 token 以及已经处理过的 token 的信息.
            # e.g. mask matrix:
            # ┌         ┐ 
            # │ 0 -∞ -∞ │
            # │ 0  0 -∞ │ 
            # │ 0  0  0 │
            # └         ┘
```

File: llama/model_cn_comment.py [hunk 14]
```diff
@@ -260,63 +270,70 @@ def forward(
...
+        values = repeat_kv(values, self.n_rep)
+
+        # [Shape] xq: (batch_size, n_local_heads, seqlen, head_dim)
+        xq = xq.transpose(1, 2)
+        # [Shape] keys: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        keys = keys.transpose(1, 2)
+        # [Shape] values: (batch_size, n_local_heads, cache_len + seqlen, head_dim)
+        values = values.transpose(1, 2)
+        # 计算 attention scores.
         scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:322]:
```suggestion
        # [Shape] scores: (batch_size, n_local_heads, seqlen, cache_len + seqlen)
        scores = torch.matmul(xq, keys.transpose(2, 3)) / math.sqrt(
```

File: llama/model_cn_comment.py [hunk 15]
```diff
@@ -260,63 +270,70 @@ def forward(
...
             self.head_dim
         )
+        # [Shape] scores: (batch_size, n_local_heads, seqlen, cache_len + seqlen)
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:325]:
```suggestion
```

File: llama/model_cn_comment.py [hunk 16]
```diff
@@ -446,20 +465,21 @@ class Transformer(nn.Module):
...
-            vocab_size (int): Vocabulary size.
-            n_layers (int): Number of layers in the model.
-            tok_embeddings (ParallelEmbedding): Token embeddings.
-            layers (torch.nn.ModuleList): List of Transformer blocks.
-            norm (RMSNorm): Layer normalization for the model output.
-            output (ColumnParallelLinear): Linear layer for final output.
-            freqs_cis (torch.Tensor): Precomputed cosine and sine frequencies.
+            params (ModelArgs): 模型配置参数.
+            vocab_size (int): 词表大小.
+            n_layers (int): 模型中的层数.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:476]:
```suggestion
            n_layers (int): 模型中 transformer block 的层数.
```

File: llama/model_cn_comment.py [hunk 17]
```diff
@@ -330,24 +347,24 @@ def __init__(
...
+            w3 (ColumnParallelLinear): 第三层的线性变换.
 
         """
         super().__init__()
         hidden_dim = int(2 * hidden_dim / 3)
-        # custom dim factor multiplier
+        # 自定义的隐藏层维度缩放参数.
         if ffn_dim_multiplier is not None:
             hidden_dim = int(ffn_dim_multiplier * hidden_dim)
         hidden_dim = multiple_of * (
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:370]:
```suggestion
        # 保证 hidden_dim 为 multiple_of 的整数倍.
        hidden_dim = multiple_of * (
```

File: llama/model_cn_comment.py [hunk 18]
```diff
@@ -260,63 +270,70 @@ def forward(
...
-
-        xq = xq.transpose(1, 2)  # (bs, n_local_heads, seqlen, head_dim)
-        keys = keys.transpose(
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
-        values = values.transpose(
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
+        # 若 n_kv_heads < n_heads, 则在 n_kv_heads 维度上进行重复扩充
+        # 使 query, key, value 的尺寸匹配.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:309]:
```suggestion
        # 使 query, key, value 的尺寸匹配.
        # Q: 为什么会出现 n_kv_heads < n_heads 的情况？
        # A: 当 transformer 层采用 Grouped-Query Attention 时会出现 n_kv_heads < n_heads 的情况.
        # 即多个 query head 共享一对 key head 和 value head, 来减小 KV-Cache, 因此 n_heads 一定能被 n_kv_heads 整除. 
```

File: llama/model_cn_comment.py [hunk 19]
```diff
@@ -423,21 +440,23 @@ def forward(
...
     ):
-        """Perform a forward pass through the TransformerBlock.
+        """执行 TransformerBlock 的前向过程.
 
         Args:
-            x (torch.Tensor): Input tensor.
-            start_pos (int): Starting position for attention caching.
-            freqs_cis (torch.Tensor): Precomputed cosine and sine frequencies.
-            mask (torch.Tensor, optional): Masking tensor for attention. Defaults to None.
+            x (torch.Tensor): 输入张量.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:446]:
```suggestion
            x (torch.Tensor): 输入张量. Shape: (batch_size, seqlen, dim)
```

File: llama/model_cn_comment.py [hunk 20]
```diff
@@ -423,21 +440,23 @@ def forward(
...
+            start_pos (int): Attention caching 的起始位置.
+            freqs_cis (torch.Tensor): 预计算的复指数频率张量.
+            mask (torch.Tensor, optional): 计算 attention 时的 mask. 默认设为 None.
 
         Returns:
-            torch.Tensor: Output tensor after applying attention and feedforward layers.
+            torch.Tensor: 经过 attention 和 FeedForward 层后输出的张量.
 
         """
+        # 残差计算 attention.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:455]:
```suggestion
        # 残差计算 attention.
        # [Shape] h: (batch_size, seqlen, dim)
```

File: llama/model_cn_comment.py [hunk 21]
```diff
@@ -423,21 +440,23 @@ def forward(
...
         h = x + self.attention(
             self.attention_norm(x), start_pos, freqs_cis, mask
         )
+        # 残差计算 FeedForward.
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:459]:
```suggestion
        # 残差计算 FeedForward.
        # [Shape] out: (batch_size, seqlen, dim)
```

File: llama/model_cn_comment.py [hunk 22]
```diff
@@ -260,63 +270,72 @@ def forward(
...
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
-        values = values.transpose(
-            1, 2
-        )  # (bs, n_local_heads, cache_len + seqlen, head_dim)
+        # 若 n_kv_heads < n_heads, 则在 n_kv_heads 维度上进行重复扩充
+        # 使 query, key, value 的尺寸匹配.
+        # Q: 为什么会出现 n_kv_heads < n_heads 的情况？
+        # A: 当 transformer 层采用 Grouped-Query Attention 时会出现 n_kv_heads < n_heads 的情况.
+        # 即多个 query head 共享一对 key head 和 value head, 来减小 KV-Cache, 因此 n_heads 一定能被 n_kv_heads 整除.
```
Comment (keli-wen) [inline llama/model_cn_comment.py:312]: Why can we say "一定"? Any proof?

File: llama/model_cn_comment.py (hunk 1 above)
Comment (keli-wen) [reply llama/model_cn_comment.py:510]: Please remember to add a [TODO] tag to refer the issue.

**Review by keli-wen — CHANGES_REQUESTED** [review]

File: llama/model_cn_comment.py [hunk 23]
```diff
@@ -480,7 +505,9 @@ def __init__(self, params: ModelArgs):
             params.dim, params.vocab_size, bias=False, init_method=lambda x: x
         )
 
+        # 计算用于 RoPE 的复指数频率张量.
         self.freqs_cis = precompute_freqs_cis(
+            # ? 暂时搁置
```
Comment (NANWOOD) [inline llama/model_cn_comment.py:510]:
```suggestion
            # ? 暂时搁置
            # [TODO] https://github.com/Mixture-AI/meta-llama-explain/issues/23#issue-2322926483
```

File: llama/model_cn_comment.py (hunk 5 above)
Comment (keli-wen) [reply llama/model_cn_comment.py:557]: How about:

```
# ┌─────────────────────────────────────────────────────────┐
# │ > mask maxtirx visualization                            │
# │                                                         │
# │                                     hstack              │
# │                                       ↓                 │
# │            ↙ [0][0][0][0][0][0][0][0] | [0][x][x][x][x] │
# │           ↙  [0][0][0][0][0][0][0][0] | [0][0][x][x][x] │
# │ seq_len(5) ← [0][0][0][0][0][0][0][0] | [0][0][0][x][x] │
# │           ↖  [0][0][0][0][0][0][0][0] | [0][0][0][0][x] │
# │            ↖ [0][0][0][0][0][0][0][0] | [0][0][0][0][0] │
# │                                    ↑     ↘  seq_len  ↙  │
# │                               cache_len(8)              │
# │                                                         │
# │ x: denote the -inf value (masked value)                 │
# └─────────────────────────────────────────────────────────┘
```

File: llama/model_cn_comment.py [hunk 24]
```diff
@@ -490,39 +517,57 @@ def __init__(self, params: ModelArgs):
...
-            # When performing key-value caching, we compute the attention scores
-            # only for the new sequence. Thus, the matrix of scores is of size
-            # (seqlen, cache_len + seqlen), and the only masked entries are (i, j) for
-            # j > cache_len + i, since row i corresponds to token cache_len + i.
+            # 由于使用了 key-value caching, 我们只需要为新序列计算 attention scores.
+            # 因此, 我们需要计算的 attention scores 的尺寸为 (seqlen, cache_len + seqlen).
+            # 那么我们需要 mask 的元素 (i, j) 当 j > cache_len + i. 实际操作时, 我们只需要
+            # 在先前计算的尺寸为 [seqlen, seqlen] 的 mask 矩阵前补上 cached 部分, 即
+            # 起始位置前的部分, 这一部分的所有元素都不需要被 mask, 所以只需要拼接上一个
+            # 尺寸为 (seqlen, start_pos) 的全零矩阵即可.
```
Comment (keli-wen) [inline llama/model_cn_comment.py:562]:
```suggestion
            # 尺寸为 (seqlen, start_pos) 的全零矩阵即可.
            # ┌─────────────────────────────────────────────────────────┐
            # │ > mask maxtirx visualization                            │
            # │                                                         │
            # │                                     hstack              │
            # │                                       ↓                 │
            # │            ↙ [0][0][0][0][0][0][0][0] | [0][x][x][x][x] │
            # │           ↙  [0][0][0][0][0][0][0][0] | [0][0][x][x][x] │
            # │ seq_len(5) ← [0][0][0][0][0][0][0][0] | [0][0][0][x][x] │
            # │           ↖  [0][0][0][0][0][0][0][0] | [0][0][0][0][x] │
            # │            ↖ [0][0][0][0][0][0][0][0] | [0][0][0][0][0] │
            # │                                    ↑     ↘  seq_len  ↙  │
            # │                               cache_len(8)              │
            # │                                                         │
            # │ x: denote the -inf value (masked value)                 │
            # └─────────────────────────────────────────────────────────┘
```

**Review by keli-wen — COMMENTED** [review]
LGTM

**Review by keli-wen — APPROVED** [review]

**Pull request merged**